### PR TITLE
Enable jitPersistentAlloc/Free to use non-global persistent memory

### PR DESCRIPTION
**Pull request by dmitry-ten**
Instead of using the global `::trPersistentMemory` object,
`jitPersistentAlloc/Free` now use a new method
`OMR::CompilerEnv::persistentMemory`.
This does not change the behavior for OMR, but upstream
projects can override `persistentMemory` method and return
memory objects that are not the global one.

### CODE REVIEW

**Comment by dmitry-ten** [comment]
Prerequisite for https://github.com/eclipse/openj9/pull/11686/

**Comment by 0xdaryl** [comment]
@dsouzai : please offer any opinions you have on this.

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
I'll merge later today unless there are any objections.

**Comment by dmitry-ten** [comment]
This probably needs to be delayed until @mpirvu reviews eclipse/openj9#11686 and says that he's ok with that implementation.
If Marius says that he doesn't want to subclass persistent memory, this commit won't be needed.

**Comment by 0xdaryl** [comment]
Please mark this PR WIP or move into Draft state to prevent it from being merged before you're ready.

**Comment by dsouzai** [comment]
Sorry for the delay in responding. Given that `TR_PersistentMemory` allocates memory using the `TR::PersistentAllocator`, why do you need to override `TR_PersistentMemory`; can't you extend `TR::PersistentAllocator`?

A slightly subtle point, though probably not that significant, is that the `TR_PersistentMemory` object gets placed in the very first persistent segment allocated from the VM; as such we put `static const uintptr_t MEMINFO_SIGNATURE = 0x1CEDD1CE;` in case we need to find it, in case we need to look at the `size_t _totalPersistentAllocations[TR_MemoryBase::NumObjectTypes];` (which we have done in the past when investigating memory leaks). If virtual methods are added, when looking in a debugger, we would need to remember that there is a VFT and adjust the address we print from appropriately.

**Comment by dmitry-ten** [comment]
> Given that TR_PersistentMemory allocates memory using the TR::PersistentAllocator, why do you need to override TR_PersistentMemory; can't you extend TR::PersistentAllocator?

I'm implementing per-client persistent allocators in eclipse/openj9#11686.
I am overriding methods of `TR_PersistentMemory` to make them automatically return a per-client allocator, when one is available. 
I also added `_totalPersistentPerClientAllocations` to my subclass of `TR_PersistentMemory` to keep track of both global and per-client allocations. 

> we put static const uintptr_t MEMINFO_SIGNATURE = 0x1CEDD1CE

I'm not sure I understand, we always put `_totalPersistentAllocations` at `MEMINFO_SIGNATURE` address?

**Comment by dsouzai** [comment]
Hm I see; yeah I guess making the methods virtual, while not ideal from a conceptual point of view, is probably still the best way forward; there are too many places that touch the `TR_PersistentMemory` object directly.

> I'm not sure I understand, we always put _totalPersistentAllocations at MEMINFO_SIGNATURE address?

No, I meant to say we use the eyecatcher to find the object in memory; once we know where the object is, we can read from the allocations array.

**Comment by dmitry-ten** [comment]
After discussing this with @mpirvu, I changed this to avoid making any new methods `virtual`. 
Instead, I modified `jitPersistentAlloc/Free` to use `CompilerEnv::persistentMemory` to get `TR_PersistentMemory` object. For OMR and regular OpenJ9 this will be the same global object as before, but JITServer will override `CompilerEnv::persistentMemory` to return a per-client instance of persistent memory. Since `CompilerEnv` is extended through extensible classes, this does not require marking anything as virtual.
@dsouzai fyi.

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
@dsouzai What is your opinion on this omr change? Thanks

**Comment by dsouzai** [comment]
The idea looks fine to me; that said, is there going to be a noticeable impact on perf for going through the TLH every time we do a `jitPersistentAlloc`? `TR_MemoryBase::jitPersistentAlloc` is part of the `TR_ALLOC` macro so there's quite a lot of places that will invoke it.

**Comment by mpirvu** [comment]
> is there going to be a noticeable impact on perf for going through the TLH

I don't think we use persistentAlloc as much as the region allocator. In the profiles that I looked at persistent allocations were never hot. That being said, we should measure start-up and rampup performance since this is where we most of the persistent allocations for IProfiler, CHTable and runtime assumptions.

This change replaces a virtual call (previous solution) with a TLH access. I think we can avoid most of the overhead if we first use a runtime test for JITServer (in OpenJ9):
```
if (JITServer) // --> this needs to be cheap, accessing some static/global variable
   auto compInfoPT = TR::compInfoPT;
   ...
else
   return OMR::CompilerEnv::persistentAllocator();
```
That first test should be well predicted by the CPU because it always returns the same value.

**Comment by dmitry-ten** [comment]
Rebased to the latest OMR.

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
@dsouzai Given the proposed OpenJ9 implementation described above in https://github.com/eclipse/omr/pull/5760#issuecomment-768500314 (implemented in https://github.com/eclipse/openj9/pull/11686) are you ok with the present OMR implementation? Thanks

@0xdaryl Could we please get a review from you (or suggest another omr committer) as well? Thanks

**Review by dsouzai — APPROVED** [review]

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**
